scarb-metadata/src/scarb_command.rs

```diff
@@ -0,0 +1,210 @@
+use std::collections::HashMap;
+use std::ffi::{OsStr, OsString};
+use std::path::PathBuf;
+use std::process::{Command, Output, Stdio};
+use std::{env, io};
+
+use thiserror::Error;
+
+/// Error thrown while trying to execute `scarb` command.
+#[derive(Error, Debug)]
+#[non_exhaustive]
+pub enum ScarbCommandError {
+    /// Failed to read `scarb` output.
+    #[error("failed to read `scarb` output")]
+    Io(#[from] io::Error),
+    /// Error during execution of `scarb` command.
+    #[error("`scarb` exited with error\n\nstdout:\n{stdout}\nstderr:\n{stderr}")]
+    ScarbError {
+        /// Captured standard output if any.
+        stdout: String,
+        /// Captured standard error if any.
+        stderr: String,
+    },
+}
+
+/// A builder for `scarb` command invocation.
+#[derive(Clone, Debug, Default)]
+pub struct ScarbCommandBuilder {
```

Comment:
I don't think anybody will want to use this to collect Scarb output, neither we should promote doing this. The only command that we expect people to read is `scarb metadata` which is covered by MetadataCommand API already. What we want here is just an ability to shell out to Scarb, always inheriting stdout and stderr.

So I think what we pursue here is something simpler, like:

```rust
ScarbCommand::new("build")
    .arg(...)
    .current_dir(...)
    .run()? -> Result<(), ScarbCommandError>
```

I know there is JSON output option everywhere, but reading it is exotic case enough I'd rather ask people to use `std::command::Command`